### DISCUSSION TITLE
Need some help with automation

### DISCUSSION THREAD

**Discussion by Nexetion** [ybizeul/StreamDeckWS#23] [General]
![image](https://user-images.githubusercontent.com/182035/155887884-0a162145-ef2c-4592-8844-4af671517e72.png)

this is my current configuration and I'm trying to figure out how make the button title change when the state changes, as of right now it changes when I press it but there are 4 states to the device I'm controlling open, close, closing, and opening. Appreciate any help!

**Comment by ybizeul** [comment]
Did you make sure that no Title is configured in StreamDeck itself ?

EDIT :

If the title is changing when you push the button, the issue is probably not my first thought.

Now, the documentation probably needs an update, in early releases back when the original version was published, it was a little bit tricky to maintain proper `context` especially with messages that weren't originated by a key push, and an update sent back to the SD (this scenario is easier to implement, as you have noticed).

If the action is initiated outside of the SD, you needed to send the correct `context` alongside with the message so the SD knows what key you're targeting, it works by sending a `context` that is generated within SD that you had to store somehow in a cache.

That's why I implemented two new nodes : __sd-input__ and __sd-output__.

Make sure you have a __sd-input__ node after the __websocket in__ node, and a __sd-output__ node before __websocket out__, they will set the flow persistent state with all the keys sending `willAppear` messages, and retore proper context before sending messages back. Then you can take events out of any source (like HA) and send it straight to __sd-output__.

I like to have a "master" block that handles __input__ and __output__, and link to task specific blocks like so :

<img width="969" alt="image" src="https://user-images.githubusercontent.com/182035/155888364-31a940ad-137a-4f6e-a807-730796e4d5fd.png">

Hope this helps.

**Reply by Nexetion** [reply]
How do I add the input? I can't find it in Node-Red/HA

**Reply by ybizeul** [reply]
That's a Streadeck-WS node that you will find under network if you're up to date.
<img width="152" alt="image" src="https://user-images.githubusercontent.com/182035/155890053-40fae67a-dbe9-4bfa-bb62-f55dfd9408d4.png">

**Reply by Nexetion** [reply]
So i would basically create this the same as how you have it in the streamdeck part of the picture?

**Reply by ybizeul** [reply]
yes, after __websocket in__ you add a __sd-input__, then you use __sd-output__ to format the message you wish to send to __websocket out__.

<img width="832" alt="image" src="https://user-images.githubusercontent.com/182035/155890290-55f13dc9-46e4-45df-b698-d001c6357640.png">

**Comment by Nexetion** [comment]
I'm either getting a string is undefined for payload.state for the title or context for temp is unknown. Anything I should screen shot to help you, help me?

**Reply by Nexetion** [reply]
Yeah I don't understand how to get the state of my garage door with this new SD input and output

**Reply by ybizeul** [reply]
__sd-input__ goes right after __websocket in__, and keeps going to where you want to catch SD events.

Specifically for what you're asking, you're not using __websocket in__ though, so it's irrelevant _except for the fact it's used to record all your keys in the flow cache_, so do it anyway.

What you need is send the message with __sd-output__ pushing into it a message, and reference the title property in the window above. Don't put anything in __State__

**Reply by Nexetion** [reply]
![image](https://user-images.githubusercontent.com/182035/155892800-c09513d5-68fc-49cc-b311-ed8a4f8e0587.png)

![image](https://user-images.githubusercontent.com/182035/155892810-a6d403f8-ea4e-4623-8624-600107e48a60.png)

**Reply by ybizeul** [reply]
Remove the `.` before `payload.state`

**Comment by Nexetion** [comment]
```
2/27/2022, 12:36:26 PM[node: 2nd Pay](http://homeassistant.local:8123/api/hassio_ingress/9rLW5uJysh3Bcxjnv_aIRZYvPrFK3cCOFqm2m7B3jU8/#)
msg.payload : string[321]
"{"action":"org.tynsoe.streamdeck.wsproxy.proxy","context":"69F862433908EC0A923E6E87A8340C86","device":"7A9B7F6073A842147791757CB8A0B46F","event":"keyDown","payload":{"coordinates":{"column":4,"row":2},"isInMultiAction":false,"settings":{"id":"button1","remoteServer":"ws://homeassistant.local:1880/endpoint/streamdeck"}}}"
2/27/2022, 12:36:26 PM[node: 2nd Pay](http://homeassistant.local:8123/api/hassio_ingress/9rLW5uJysh3Bcxjnv_aIRZYvPrFK3cCOFqm2m7B3jU8/#)msg.payload : string[319]
"{"action":"org.tynsoe.streamdeck.wsproxy.proxy","context":"69F862433908EC0A923E6E87A8340C86","device":"7A9B7F6073A842147791757CB8A0B46F","event":"keyUp","payload":{"coordinates":{"column":4,"row":2},"isInMultiAction":false,"settings":{"id":"button1","remoteServer":"ws://homeassistant.local:1880/endpoint/streamdeck"}}}"
2/27/2022, 12:36:26 PM[node: Stream Deck](http://homeassistant.local:8123/api/hassio_ingress/9rLW5uJysh3Bcxjnv_aIRZYvPrFK3cCOFqm2m7B3jU8/#)msg : undefined
undefined
```

**Reply by Nexetion** [reply]
now i get undefined for msg.payload.state

**Reply by ybizeul** [reply]
Did you actually pu something in `msg.payload.state` ?

**Reply by Nexetion** [reply]
check comment below

**Comment by Nexetion** [comment]
![image](https://user-images.githubusercontent.com/182035/155893271-6b92be64-3da9-4187-ab1d-3a05976352b5.png)

```
2/27/2022, 12:39:43 PM[node: Stream Deck](http://homeassistant.local:8123/api/hassio_ingress/9rLW5uJysh3Bcxjnv_aIRZYvPrFK3cCOFqm2m7B3jU8/#)
msg : undefined
undefined
2/27/2022, 12:39:43 PM[node: 3rd Pay](http://homeassistant.local:8123/api/hassio_ingress/9rLW5uJysh3Bcxjnv_aIRZYvPrFK3cCOFqm2m7B3jU8/#)msg.payload.state : string[6]
"closed"
```

**Reply by ybizeul** [reply]
Can you show me the output of the HA node right before __sd-output__ ?

**Reply by Nexetion** [reply]
```
2/27/2022, 12:43:30 PM[node: Stream Deck](http://homeassistant.local:8123/api/hassio_ingress/9rLW5uJysh3Bcxjnv_aIRZYvPrFK3cCOFqm2m7B3jU8/#)
msg : undefined
undefined
2/27/2022, 12:43:30 PM[node: 4th pay](http://homeassistant.local:8123/api/hassio_ingress/9rLW5uJysh3Bcxjnv_aIRZYvPrFK3cCOFqm2m7B3jU8/#)msg.payload.state : string[6]
"closed"
2/27/2022, 12:43:30 PM[node: 3rd Pay](http://homeassistant.local:8123/api/hassio_ingress/9rLW5uJysh3Bcxjnv_aIRZYvPrFK3cCOFqm2m7B3jU8/#)msg.payload.state : string[6]
"closed"
```

**Reply by Nexetion** [reply]
it's like its going to the websocket before it can populate the state.

**Reply by ybizeul** [reply]
That's because it's getting unexpected input. Please show what's coming from the HA node

**Reply by Nexetion** [reply]
![image](https://user-images.githubusercontent.com/182035/155893633-3ba46a07-38af-4571-b3fe-555315e62306.png)

**Reply by ybizeul** [reply]
I need to see the message that is coming out of this

**Reply by Nexetion** [reply]
I FIGURED IT OUT JUST SET PAYLOAD TO THE STATE INSTEAD OF PAYLOAD.STATE

**Reply by ybizeul** [reply]
Depends on what you send yes. You could keep `entity` as the payload, but then in the __sd-output__ you need to fetch the property in that json that represents the status. probably just `state`

**Reply by Nexetion** [reply]
Now the question is how do I make the title changed based on the status instead of a keypush.

**Reply by ybizeul** [reply]
Just start with a HA __event: state__ node and push new state to the __sd-output__ node.
That __event: state__ node would be right below the __current state__ node you already have and probably go into the same __sd-output__

**Reply by Nexetion** [reply]
in the state changed how would I sent the context?

**Reply by Nexetion** [reply]
![image](https://user-images.githubusercontent.com/182035/155897630-e690a55e-4f9d-4a57-99c4-9d350e7ff357.png)

I have it like this right now.

**Reply by ybizeul** [reply]
> in the state changed how would I sent the context?

You don't have to, that's the beauty of it :-) As long as they key is "known" from a previous message it'll remember it. A previous message would be the _willAppear_ sent when the key was displayed initially. That's the main purpose of having a __sd input__ node in the flow.

**Comment by Nexetion** [comment]
Whats in the "Music"?

**Reply by ybizeul** [reply]
The _Music_ switch just checks the value of `msg.payload.payload.settings.id` and make sure that's the key with the `music` ID that has been pushed.

**Reply by Nexetion** [reply]
So how would I change the state on the key when say the garage is opened from my phone or the app on my phone with the state change node correct?

**Reply by ybizeul** [reply]
Yes, you use an __event: state__ node on the garage door entity, and that goes to a __sd-output__ node where you put the property of the message that stores the door state in the __Title__ field. You can also  use a __change__ or __function__ node between to reformat if needed

**Comment by Nexetion** [comment]
Is there anyway we could do this on discord or you could send me your flow for this somehow?

**Reply by ybizeul** [reply]
The flow for what ?

**Comment by meehab** [comment]
I appreciate I'm digging up an old thread here, but I'm struggling a little with this.

@ybizeul  It would be really useful to have the flows for the image posted on 27th Feb showing what's referred to as  master block and the task specific blocks. Would really help to be able to explore those and click into the nodes rather than just looking at the image.
